pkg/volumeclaimbinder/persistent_volume_claim_binder.go

```diff
@@ -0,0 +1,301 @@
+		return true
+	}
+	return false
+}
+
+func syncVolume(binderClient binderClient, volume *api.PersistentVolume) (err error) {
+	glog.V(5).Infof("Synchronizing PersistentVolume[%s]\n", volume.Name)
+
+	if volume.Spec.ClaimRef != nil {
+		if volume.Status.Phase == api.VolumeAvailable {
+			volume.Status.Phase = api.VolumeBound
+			_, err := binderClient.UpdatePersistentVolumeStatus(volume)
+			if err != nil {
+				return fmt.Errorf("Error updating pv.status: %v\n", err)
+			}
+		}
+
+		// verify the volume is still claimed by a user
+		if claim, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name); err == nil {
+			glog.V(5).Infof("PersistentVolume[%s] is bound to PersistentVolumeClaim[%s]\n", volume.Name, volume.Spec.ClaimRef.Name)
+			// rebuild the Claim's Status as needed
+			if claim.Status.VolumeRef == nil {
+				syncClaimStatus(binderClient, volume, claim)
+			}
+		} else {
+			//claim was deleted by user.
+			glog.V(3).Infof("PersistentVolumeClaim[%s] unbound from PersistentVolume[%s]\n", volume.Spec.ClaimRef.Name, volume.Name)
+			// volume.Spec.ClaimRef is deliberately left non-nil so that another process can recycle the newly released volume
+			volume.Status.Phase = api.VolumeReleased
+			volume, err = binderClient.UpdatePersistentVolumeStatus(volume)
+			if err != nil {
+				return fmt.Errorf("Error updating pv: %+v\n", err)
+			}
+		}
+	} else {
+		volume.Status.Phase = api.VolumeAvailable
+		_, err := binderClient.UpdatePersistentVolumeStatus(volume)
+		if err != nil {
+			return fmt.Errorf("Error updating pv.status: %v\n", err)
+		}
+	}
+	return nil
+}
+
+func syncClaimStatus(binderClient binderClient, volume *api.PersistentVolume, claim *api.PersistentVolumeClaim) (err error) {
+	volumeRef, err := api.GetReference(volume)
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting volume reference: %v\n", err)
+	}
+	// all "actuals" are transferred from PV to PVC so the user knows what
+	// type of volume they actually got for their claim
+	claim.Status.Phase = api.ClaimBound
+	claim.Status.VolumeRef = volumeRef
+	claim.Status.AccessModes = volume.Spec.AccessModes
+	claim.Status.Capacity = volume.Spec.Capacity
+
+	_, err = binderClient.UpdatePersistentVolumeClaimStatus(claim)
+
+	if err != nil {
+		claim.Status.Phase = api.ClaimPending
+		claim.Status.VolumeRef = nil
+		claim.Status.AccessModes = nil
+		claim.Status.Capacity = nil
+	}
+
+	return err
+}
+
+func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderClient, claim *api.PersistentVolumeClaim) (err error) {
+	glog.V(5).Infof("Synchronizing PersistentVolumeClaim[%s]\n", claim.Name)
+
+	if claim.Status.VolumeRef != nil {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] is bound to PersistentVolume[%s]\n", claim.Name, claim.Status.VolumeRef.Name)
+		return nil
+	}
+
+	volume, err := volumeIndex.FindBestMatchForClaim(claim)
+	if err != nil {
+		return err
+	}
+
+	if volume != nil {
```

Comment:
flip this to be if volume == nil first, handle error and return.  undent the rest.